pywhy_graphs/algorithms/generic.py

```diff
@@ -333,3 +334,207 @@ def _single_shortest_path_early_stop(G, firstlevel, paths, cutoff, join, valid_p
                     nextlevel[w] = 1
         level += 1
     return paths
+
+
+def _find_directed_parents(G, node):
+    """Finds the parents of a node in the directed and the bidirected subgraphs.
```

Comment:
This is a weird name for the function. When I review it, it sounds like it should only find the parents in the directed subgraph of `node`, but in fact it finds the parents in the directed subgraph + the adjacent neighbors.

To make things more maintainable, it is better to have a separate function for:

1. parents in the directed subgraph
2. neighbors in the bidirected subgraph